### PR TITLE
Fixes packager to ensure the GUIDs are used for all entities where possible

### PR DESCRIPTION
**Pull request by Shazwazza**
When package data is installed we want to make sure that it installs using the GUIDs assigned/exported so that new entities are not created when the package is reinstalled. This is the case for many entities but for others it isn't.

* DataTypes = OK
* Languages = By ISO Code = OK (no GUID in the DB)
* Dictionary = NOT OK, only looked up by the string key but needs to use the GUID instead
* Macros = NOT OK, only looked up by string alias but needs to use the GUID
* Template = OK, looked up by alias because there is no GUID in the DB
* Doc types = NOT OK, only looked up by alias but needs to use the GUID
* Media types = NOT OK, only looked up by alias but needs to use the GUID
* Stylesheets = OK, uses the name, this is just a file anyways
* Documents = OK, uses the 'key' (GUID) but it seems to be optional so lets make that not optional (see CreateContentFromXml)
* Media = OK, uses the 'key' (GUID) but it seems to be optional so lets make that not optional (see CreateContentFromXml)
* Macro properties = NOT OK
* Property types = NOT OK

## Testing

This is a little annoying but here's what to do:

* Create a new doc type with one property
* Create a 2nd nested doc type with one property (to ensure that the nesting still works with GUIDs)
* Create a media type with one property
* Create a couple nested dictionary items
* Create a macro with a property
* Then create a package with all of these specific items selected
* Then export the package.xml file
* Then go delete all of these created items in Umbraco

In your webhost testing project (i.e. `Umbraco.Web.UI.NetCore`), Create a folder/namespace like `MyPackage`. Then add an automatic package migration there like:

```cs
public class TestAutomaticPackageMigrationPlan : AutomaticPackageMigrationPlan
{
    public TestAutomaticPackageMigrationPlan() : base("MyPackage")
    {
    }
}
```

The move your downloaded package.xml file into this folder and mark as an embedded resource.

Then start your project. This will run the automatic package migration and install all of this data. You'll then need to go into the DB to lookup the items created in the various tables like umbracoNode (for doc/media types), cmsPropertyType (for property types), cmsMacro (for macros), cmsMacroProperty (for macro properties), cmsLanguageText (for dictionary items). And then verify all of the GUIDs are the same as all of the GUIDs in your package.xml file.

Then stop the site and make a small change to the file, like removing an empty `<Validation />` node. This will change the hash of this file which is used to create the automatic package migration plan Id. So then start the site again and it will run another package migration for the same file. Then you can verify that it didn't make any duplicates and all the GUIDs are still the same.

### CODE REVIEW

**Review by bergmania — COMMENTED** [review]
Code changes looks fine.. I still need to test. But also we one test failing, which could be related to the code, and not just tests that needed to be updated (I updated all the embedded xml used in tests, and added Guids).

**Comment by bergmania** [comment]
In my test, most worked as expected, but also here it looks like the dictionary items have problems (maybe related to the failing test)..

What happened is that I created a package with the following dictionary item structure

- Level 1
  - Level 2
     - Level 3
     
But the dictionary items are restored as:
- Level 1
- Level 2
- Level 3

This seems to be related to the packing of them, as the xml looks like this in `package.xml`:
```xml
  <DictionaryItems>
    <DictionaryItem Id="5e1ec2c5-f06e-49cc-a853-6f0acfb0d957" Key="Level 1" />
    <DictionaryItem Id="a4bc5958-b16e-4fb1-adb9-ffe7c81b8e54" Key="Level 3" />
    <DictionaryItem Id="b66b7422-2598-4e4d-9a5e-cbf2f88a496d" Key="Level 2" />
  </DictionaryItems>
```

File: src/Umbraco.Infrastructure/Services/Implement/EntityXmlSerializer.cs [hunk 1]
```diff
@@ -431,6 +427,7 @@ public XElement Serialize(IMacro macro)
             foreach (var property in macro.Properties)
             {
                 properties.Add(new XElement("property",
+                    new XAttribute("id", property.Key),
```
[the remark this answers is not in the document]
Comment (Shazwazza) [reply src/Umbraco.Infrastructure/Services/Implement/EntityXmlSerializer.cs:430]: Thought of that too. Happy to change but for Dictionary, 'Key' is actually something different 🤦 I'll fix up the Dictionary one to be consistent too.

**Comment by Shazwazza** [comment]
> This seems to be related to the packing of them, as the xml looks like this in package.xml:

yep that's the issue. I'll have a look

**Comment by bergmania** [comment]
> > This seems to be related to the packing of them, as the xml looks like this in package.xml:
> 
> yep that's the issue. I'll have a look

Seems to have been like this for years

**Comment by Shazwazza** [comment]
> Seems to have been like this for years

Yep. I'll add some TODOs in the code and we can create a different task.

**Comment by Shazwazza** [comment]
Actually, decided to just fix it :P

**Comment by Shazwazza** [comment]
Should be good now. The packager will put the dictionary items in the xml hierarchy where possible.

**Pull request merged**